### DISCUSSION TITLE
Multitenancy Support for Eclipse Dirigible

### DISCUSSION THREAD

**Discussion by delchev** [eclipse/dirigible#3423] [Ideas]
## Overview

Eclipse Dirigible currently lacks native support for multitenancy, limiting its ability to cater to scenarios where multiple tenants need to coexist within the same instance. This feature request aims to address this gap and enhance the platform's capabilities by introducing robust multitenancy support.

## Motivation

Multitenancy is becoming increasingly essential in cloud-based applications, enabling multiple users or organizations (tenants) to securely share the same resources while maintaining isolation. Introducing multitenancy support in Eclipse Dirigible would significantly broaden its appeal and applicability in various deployment scenarios.

## Desired Features

The proposed multitenancy support should include, but not be limited to:

* `Isolation`: Ensure secure separation of data, metadata, and runtime resources between different tenants.
* `Tenant Configuration`: Allow administrators to configure and manage tenants efficiently, including setting resource limits, access controls, and specific configurations for each tenant.
* `Authentication and Authorization`: Implement mechanisms to handle authentication and authorization for each tenant independently, ensuring that users from one tenant cannot access resources of another tenant.
* `Scalability`: Design the multitenancy support to scale effectively, accommodating a growing number of tenants without compromising performance.

## Use Cases

* `SaaS Applications`: Eclipse Dirigible can be used to develop Software as a Service (SaaS) applications where different organizations or users act as separate tenants.
* `Enterprise Deployment`: Large enterprises with multiple departments or subsidiaries can benefit from a single Dirigible instance that supports multitenancy, streamlining development and resource management.

### Technology Stack

Consider leveraging industry-standard practices and technologies for implementing multitenancy, such as database schema separation, containerization, and security best practices.

### Documentation

Ensure comprehensive documentation is provided to guide users and administrators through the setup and management of multitenancy in Eclipse Dirigible.

**Comment by delchev** [comment]
### In scope (Multi-tenant):

* Security
    * User Management
    * Authentication
    * Authorization
* Database
    * Schema per Tenant 
* Jobs
    * ThreadLocal + Execution of the Handler per Tenant
* Listeners
    * Tenant prefix for Queues/Topics
* Documents
    * ThreadLocal + Tenant prefix for Document Path

### Out of scope (Single-tenant):

* Processes (BPM Flowable)
* Integrations (Apache Camel)
* Commands
* FTP/SFTP
* OData
* Web
* Wiki
* WebSockets

**Comment by iliyan-velichkov** [comment]
Implementation and details in 
- phase 1 - https://github.com/eclipse/dirigible/pull/3553
- phase 2 - https://github.com/eclipse/dirigible/pull/3588
- phase 3 - https://github.com/eclipse/dirigible/pull/3609
- phase 4 - https://github.com/eclipse/dirigible/pull/3613
